src/storage/petastorm_storage_engine.py

```diff
@@ -0,0 +1,137 @@
+from src.catalog.models.df_metadata import DataFrameMetadata
+from petastorm.etl.dataset_metadata import materialize_dataset
+from src.storage.abstract_storage_engine import AbstractStorageEngine
+from src.configuration.configuration_manager import ConfigurationManager
+
+from petastorm.unischema import dict_to_spark_row
+from petastorm.predicates import in_set, in_lambda
+from petastorm import make_reader
+from typing import Iterator, Dict
+from pathlib import Path
+
+
+class PetastormStorageEngine(AbstractStorageEngine):
+
+    def __init__(self):
+        """
+        Maintain a long live spark session and context.
+        """
+        self._spark = Session()
+        self.spark_session = self._spark.get_session()
+        self.spark_context = self._spark.get_context()
+
+    def _spark_url(self, table: DataFrameMetadata) -> str:
+        """
+        Generate a spark/petastorm url given a table
+        """
+        eva_dir = ConfigurationManager().get_value("core", "location")
+        output_url = os.path.join(eva_dir, table.name)
+
+        return Path(output_url).resolve().as_uri()
+
+    def create(self, table: DataFrameMetadata):
+        """
+        Create an empty dataframe in petastorm.
+        """
+        empty_rdd = self.spark_context.emptyRDD()
+
+        with materialize_dataset(self.spark_session,
+                                 self._spark_url(table),
+                                 table.schema.petastorm_schema):
+
+            self.spark_session.createDataFrame(empty_rdd,
+                                               table.schema.pyspark_schema) \
+                .coalesce(1) \
+                .write \
+                .mode('overwrite') \
+                .parquet(self._spark_url(table))
+
+    def write_row(self, table: DataFrameMetadata, rows: []):
+        """
+        Write rows into the dataframe.
+
+        Arguments:
+            rows List[Dict{key, value}]: We can formally define a dataType Row.
+            keys within the Dict should be consistent with the table.schema.
+        """
+        with materialize_dataset(self.spark_session,
+                                 self._spark_url(table),
+                                 table.schema.petastorm_schema):
+
+            rows_rdd = self.spark_context.parallelize(rows) \
+                .map(lambda x: dict_to_spark_row(table.schema.petastorm_schema,
+                                                 x))
+
+            self.spark_session.createDataFrame(rows_rdd,
+                                               table.schema.pyspark_schema) \
+                .coalesce(1) \
+                .write \
+                .mode('append') \
+                .parquet(self._spark_url(table))
+
+    def read(self, table: DataFrameMetadata) -> Iterator[Dict]:
+        """
+        Read the dataframe.
+
+        Return:
+            Iterator of Dict. Each Dict represents a Row.
+        """
+        with make_reader(self._spark_url(table)) as reader:
```

Comment:
We are accidentally creating a new file for every read. Actually the table object should have the file path, we need to convert that to uri before feeding to the petastorm reader.

Reply:
We have the petastorm reader. I think we should use that class here instead of make_reader. Maybe add the predicate parameter support in the Petastorm reader since we need that for lambda.

Reply:
Quick question (probably, I don't have the context): Where are we taking care of the `shard` we need to read?

Reply:
We need to implement a storage context to take care of the sharding. Probably delaying that logic to late PR. Right now assuming we have a single server.